File: ios/Classes/PaymentStreamHandler.swift
//
//  PaymentStreamHandler.swift
//  flutter_klarna_payment
//
//  Created by qaq on 08/10/2023.
//

import Flutter
import KlarnaMobileSDK

class PaymentStreamHandler: NSObject, FlutterStreamHandler {
    private var paymentView: KlarnaPaymentView?
    private var sink: FlutterEventSink?

    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        sink = events
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        sink = nil
        return nil
    }

    func setPaymentView(_ pv: KlarnaPaymentView) {
        paymentView = pv
    }

    func sendMessage(state: String, message: String?) {
        sink?(["state": state, "message": message ?? ""])
    }

    func pay() {
        if let paymentView = paymentView {
            paymentView.authorize()
        }
    }
}

File: android/src/main/kotlin/com/flutter_klarna_payment/flutter_klarna_payment/PaymentStreamHandler.kt
package com.flutter_klarna_payment.flutter_klarna_payment

import com.klarna.mobile.sdk.api.payments.KlarnaPaymentView
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.EventChannel.EventSink
import io.flutter.plugin.common.EventChannel.StreamHandler

class PaymentStreamHandler() : StreamHandler {

    private var paymentView: KlarnaPaymentView? = null;
    private var sink : EventSink? = null

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        sink = events
        sendMessage(KlarnaPaymentState.INITIAL.toCamelCase(),null)
    }

    override fun onCancel(arguments: Any?) {
        sink = null;
    }

    fun setPaymentView(pv:KlarnaPaymentView){
        paymentView=pv;
    }

    fun sendMessage(state:String, message:String?){
        sink?.success(mapOf("state" to state, "message" to message));
    }

    fun pay(){
        if (paymentView!=null){
            paymentView!!.authorize(false,null)
        }
    }
}